File: engine/drift_policy.py
"""Consumer-owned drift/projection policy."""
import json
import re

from engine import paths


class DriftPolicyError(ValueError):
    pass


_NAME_RE = re.compile(r"[A-Za-z_][A-Za-z0-9_]*")
_TOP_LEVEL_KEYS = frozenset(("version", "resource_types"))
_RESOURCE_KEYS = frozenset((
    "projection_omit",
    "projection_sync",
    "projection_omit_if",
    "plan_tolerate",
))
_COMMON_ENTRY_KEYS = frozenset(("path", "reason", "approved_by", "ticket"))
_PROJECTION_SYNC_ENTRY_KEYS = frozenset((
    "target_path", "source_path", "reason", "approved_by", "ticket"
))
_PROJECTION_OMIT_IF_ENTRY_KEYS = (
    _COMMON_ENTRY_KEYS | frozenset(("values",))
)
_PLAN_TOLERATE_ENTRY_KEYS = _COMMON_ENTRY_KEYS | frozenset(("actions",))
_SAFE_PLAN_ACTIONS = frozenset(("update",))
_DEFAULT_STALE_MODES = (
    "projection_omit",
    "projection_sync",
    "projection_omit_if",
    "plan_tolerate",
)


class DriftPolicy(object):
    def __init__(self, data, source="<memory>"):
        self.data = (
            {"version": 1, "resource_types": {}}
            if data is None else data
        )
        self.source = source
        self._validate()
        self._matched_ids = set()

    @classmethod
    def load(cls, path):
        if not path:
            return cls(None)
        with open(path, encoding="utf-8") as f:
            return cls(json.load(f), source=path)

    def projection_omits(self, resource_type, path_tuple):
        return self._matches(resource_type, "projection_omit", path_tuple)

    def tolerates_plan_path(self, resource_type, path_tuple, action):
        for entry in self._entries(resource_type, "plan_tolerate"):
            if action not in entry.get("actions", ["update"]):
                continue
            if paths.selector_matches(parse_path(entry["path"]), path_tuple):
                self._matched_ids.add(id(entry))
                return True
        return False

    def stale_entries(self, resource_types=None, modes=None):
        resource_types = set(resource_types or [])
        modes = tuple(modes or _DEFAULT_STALE_MODES)
        stale = []
        for rt, cfg in sorted((self.data.get("resource_types") or {}).items()):
            if resource_types and rt not in resource_types:
                continue
            for mode in modes:
                for entry in cfg.get(mode) or []:
                    if id(entry) not in self._matched_ids:
                        stale.append((rt, mode, _entry_display_path(entry)))
        return stale

    def _matches(self, resource_type, mode, path_tuple):
        for entry in self._entries(resource_type, mode):
            if paths.selector_matches(parse_path(entry["path"]), path_tuple):
                self._matched_ids.add(id(entry))
                return True
        return False

    def entries(self, resource_type, mode):
        """Public read accessor for policy entries. Do not mutate the result."""
        return list(self._entries(resource_type, mode))

    def mark_matched(self, entry):
        self._matched_ids.add(id(entry))

    def _entries(self, resource_type, mode):
        return (
            (self.data.get("resource_types") or {})
            .get(resource_type, {})
            .get(mode, [])
        )

    def _validate(self):
        if not isinstance(self.data, dict):
            raise DriftPolicyError("%s: drift policy must be an object" % self.source)
        _reject_unknown_keys(
            self.data, _TOP_LEVEL_KEYS, "%s top-level drift policy" % self.source
        )
        if "version" not in self.data:
            raise DriftPolicyError("%s: drift policy missing version" % self.source)
        if self.data.get("version") != 1:
            raise DriftPolicyError(
                "%s: unsupported drift policy version" % self.source
            )
        if "resource_types" not in self.data:
            raise DriftPolicyError(
                "%s: drift policy missing resource_types" % self.source
            )
        if not isinstance(self.data.get("resource_types"), dict):
            raise DriftPolicyError("%s: resource_types must be an object" % self.source)
        for rt, cfg in sorted(self.data.get("resource_types").items()):
            if not isinstance(rt, str) or not _NAME_RE.fullmatch(rt):
                raise DriftPolicyError(
                    "%s: invalid resource type %r" % (self.source, rt)
                )
            if not isinstance(cfg, dict):
                raise DriftPolicyError(
                    "%s: policy for %s must be an object" % (self.source, rt)
                )
            _reject_unknown_keys(
                cfg, _RESOURCE_KEYS, "%s policy for %s" % (self.source, rt)
            )
            for mode in _DEFAULT_STALE_MODES:
                entries = cfg.get(mode, [])
                if not isinstance(entries, list):
                    raise DriftPolicyError(
                        "%s %s entries for %s must be a list"
                        % (self.source, mode, rt)
                    )
                seen = {}
                for entry in entries:
                    path, scope = self._validate_entry(rt, mode, entry)
                    prior = seen.get(scope)
                    if prior is not None:
                        raise DriftPolicyError(
                            "%s duplicate %s entry for %s path %s"
                            % (self.source, mode, rt, path)
                        )
                    seen[scope] = entry

    def _validate_entry(self, rt, mode, entry):
        if not isinstance(entry, dict):
            raise DriftPolicyError(
                "%s %s entry for %s must be an object" % (self.source, mode, rt)
            )
        allowed = _entry_keys_for_mode(mode)
        _reject_unknown_keys(
            entry,
            allowed,
            "%s %s entry for %s" % (self.source, mode, rt),
        )
        required_fields = (
            ("target_path", "source_path", "reason", "approved_by")
            if mode == "projection_sync" else
            ("path", "values", "reason", "approved_by")
            if mode == "projection_omit_if" else
            ("path", "reason", "approved_by")
        )
        for required in required_fields:
            if required == "values":
                continue
            if not isinstance(entry.get(required), str) or not entry.get(required):
                raise DriftPolicyError(
                    "%s %s entry for %s missing %s"
                    % (self.source, mode, rt, required)
                )
        if "ticket" in entry and (
                not isinstance(entry.get("ticket"), str) or not entry.get("ticket")):
            raise DriftPolicyError(
                "%s %s entry for %s has invalid ticket"
                % (self.source, mode, rt)
            )
        if mode == "projection_sync":
            target_path = entry["target_path"]
            source_path = entry["source_path"]
            target = parse_path(target_path)
            source = parse_path(source_path)
            if target == source:
                raise DriftPolicyError(
                    "%s projection_sync entry for %s target_path and "
                    "source_path must differ" % (self.source, rt)
                )
            if _has_wildcard_or_index(target):
                raise DriftPolicyError(
                    "%s projection_sync entry for %s target_path must not "
                    "contain wildcard or index selectors" % (self.source, rt)
                )
            if _has_wildcard_or_index(source):
                raise DriftPolicyError(
                    "%s projection_sync entry for %s source_path must not "
                    "contain wildcard or index selectors" % (self.source, rt)
                )
            return target_path, ("projection_sync", target_path)

        path = entry["path"]
        parse_path(path)
        if mode == "projection_omit_if":
            values = entry.get("values")
            if not isinstance(values, list) or not values:
                raise DriftPolicyError(
                    "%s projection_omit_if entry for %s values must be a "
                    "non-empty JSON list" % (self.source, rt)
                )
            for value in values:
                if not _is_json_scalar(value):
                    raise DriftPolicyError(
                        "%s projection_omit_if entry for %s values must "
                        "contain only JSON scalars" % (self.source, rt)
                    )
            return path, (
                "projection_omit_if",
                path,
                tuple(_json_scalar_key(value) for value in values),
            )
        if mode != "plan_tolerate":
            return path, ("projection_omit", path)
        actions = entry.get("actions", ["update"])
        if not isinstance(actions, list):
            raise DriftPolicyError(
                "%s plan_tolerate entry for %s actions must be a list"
                % (self.source, rt)
            )
        if not actions:
            raise DriftPolicyError(
                "%s plan_tolerate entry for %s actions must not be empty"
                % (self.source, rt)
            )
        seen_actions = set()
        for action in actions:
            if not isinstance(action, str) or not action:
                raise DriftPolicyError(
                    "%s plan_tolerate entry for %s has invalid action"
                    % (self.source, rt)
                )
            if action not in _SAFE_PLAN_ACTIONS:
                raise DriftPolicyError(
                    "%s plan_tolerate entry for %s has unsupported action %r"
                    % (self.source, rt, action)
                )
            if action in seen_actions:
                raise DriftPolicyError(
                    "%s plan_tolerate entry for %s has duplicate action %r"
                    % (self.source, rt, action)
                )
            seen_actions.add(action)
        return path, ("plan_tolerate", path, tuple(sorted(actions)))


def _entry_keys_for_mode(mode):
    if mode == "plan_tolerate":
        return _PLAN_TOLERATE_ENTRY_KEYS
    if mode == "projection_sync":
        return _PROJECTION_SYNC_ENTRY_KEYS
    if mode == "projection_omit_if":
        return _PROJECTION_OMIT_IF_ENTRY_KEYS
    return _COMMON_ENTRY_KEYS


def _entry_display_path(entry):
    return entry.get("path", entry.get("target_path"))


def _has_wildcard_or_index(path):
    return any(
        isinstance(segment, int)
        or (type(segment) is str and segment == paths.WILDCARD)
        for segment in path
    )


def _is_json_scalar(value):
    return value is None or type(value) in (str, int, float, bool)


def _json_scalar_key(value):
    if value is None:
        return ("null", None)
    if isinstance(value, bool):
        return ("bool", value)
    if isinstance(value, (int, float)):
        return ("number", value)
    return ("string", value)


def _reject_unknown_keys(obj, allowed, where):
    unknown = sorted((key for key in obj if key not in allowed), key=str)
    if unknown:
        raise DriftPolicyError("%s has unknown key %s" % (where, unknown[0]))


def parse_path(text):
    if not isinstance(text, str) or not text:
        raise DriftPolicyError("policy path must be a non-empty string")
    try:
        return paths.parse_path(text, strict_names=True, what="policy path")
    except ValueError as exc:
        raise DriftPolicyError(str(exc))

File: node-src/domain/drift-policy.ts
import { LosslessNumber } from "lossless-json";

import {
  parsePolicyPath,
  POLICY_WILDCARD,
  PolicyPathError,
  policyPathHasWildcardOrIndex,
  policyPathsEqual,
  policySelectorMatches,
  type ConcretePathSegment,
} from "./policy-paths.js";
import { terraformJsonExactlyEqual } from "../json/python-equality.js";
import { sortedStrings } from "../json/python-compatible.js";

const TOP_LEVEL_KEYS = new Set(["version", "resource_types"]);
const RESOURCE_KEYS = new Set([
  "projection_omit",
  "projection_sync",
  "projection_fill",
  "projection_omit_if",
  "plan_tolerate",
]);
const COMMON_KEYS = new Set(["path", "reason", "approved_by", "ticket"]);
const MODES = [
  "projection_omit",
  "projection_sync",
  "projection_fill",
  "projection_omit_if",
  "plan_tolerate",
] as const;
type PolicyMode = typeof MODES[number];
export type PolicyEntry = Record<string, unknown>;

const MAX_POLICY_ENTRIES = 50_000;
const MAX_PLAN_TOLERATE_WILDCARDS_PER_RESOURCE = 1_000;

interface CompiledPlanTolerate {
  readonly entry: PolicyEntry;
  readonly order: number;
  readonly selector: readonly (string | number | bigint)[];
}

export class DriftPolicyError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "DriftPolicyError";
  }
}

function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === "object" && value !== null && !Array.isArray(value);
}

function fail(message: string): never {
  throw new DriftPolicyError(message);
}

/** Accept only the exact JSON-number value one, including lossless spellings. */
export function isSupportedDriftPolicyVersion(value: unknown): boolean {
  return terraformJsonExactlyEqual(value, 1);
}

function rejectUnknownKeys(
  object: Record<string, unknown>,
  allowed: ReadonlySet<string>,
  where: string,
): void {
  const unknown = sortedStrings(Object.keys(object).filter((key) => !allowed.has(key)));
  if (unknown.length > 0) {
    fail(`${where} has unknown key ${unknown[0]}`);
  }
}

function entriesFor(
  data: Record<string, unknown>,
  resourceType: string,
  mode: PolicyMode,
): PolicyEntry[] {
  const resources = data.resource_types;
  if (!isRecord(resources)) {
    return [];
  }
  const resource = resources[resourceType];
  if (!isRecord(resource)) {
    return [];
  }
  const entries = resource[mode];
  return Array.isArray(entries) ? entries as PolicyEntry[] : [];
}

function entryKeys(mode: PolicyMode): Set<string> {
  if (mode === "projection_sync") {
    return new Set(["target_path", "source_path", "reason", "approved_by", "ticket"]);
  }
  if (mode === "projection_fill") {
    return new Set(["path", "source", "reason", "approved_by", "ticket"]);
  }
  if (mode === "projection_omit_if") {
    return new Set([...COMMON_KEYS, "values"]);
  }
  if (mode === "plan_tolerate") {
    return new Set([...COMMON_KEYS, "actions"]);
  }
  return COMMON_KEYS;
}

function requiredStrings(mode: PolicyMode): readonly string[] {
  if (mode === "projection_sync") {
    return ["target_path", "source_path", "reason", "approved_by"];
  }
  if (mode === "projection_fill") {
    return ["path", "source", "reason", "approved_by"];
  }
  return ["path", "reason", "approved_by"];
}

function requireString(entry: PolicyEntry, key: string, context: string): string {
  const value = entry[key];
  if (typeof value !== "string" || value.length === 0) {
    return fail(`${context} missing ${key}`);
  }
  return value;
}

function isJsonScalar(value: unknown): boolean {
  return value === null
    || typeof value === "string"
    || (typeof value === "number" && Number.isFinite(value))
    || typeof value === "boolean"
    || value instanceof LosslessNumber;
}

function numericScalarMarker(value: number | LosslessNumber): string {
  if (value instanceof LosslessNumber) {
    const token = value.toString();
    if (/^-?(?:0|[1-9][0-9]*)$/u.test(token)) {
      return `integer:${String(BigInt(token))}`;
    }
    const numeric = Number(token);
    return Number.isFinite(numeric) && Number.isInteger(numeric)
      ? `integer:${String(BigInt(numeric))}`
      : `float:${String(numeric)}`;
  }
  if (Number.isSafeInteger(value) && !Object.is(value, -0)) {
    return `integer:${String(BigInt(value))}`;
  }
  return Number.isInteger(value)
    ? `integer:${String(BigInt(value))}`
    : `float:${String(value)}`;
}

function jsonScalarMarker(value: unknown): string {
  if (value === null) return "null";
  if (typeof value === "boolean") return `boolean:${String(value)}`;
  if (typeof value === "string") return `string:${JSON.stringify(value)}`;
  if (typeof value === "number" || value instanceof LosslessNumber) {
    return `number:${numericScalarMarker(value)}`;
  }
  return fail("drift policy scalar marker received a non-scalar value");
}

function policyPath(value: unknown) {
  try {
    return parsePolicyPath(value);
  } catch (error: unknown) {
    if (error instanceof PolicyPathError) {
      return fail(error.message);
    }
    throw error;
  }
}

function pathMarker(path: readonly (string | number | bigint)[]): string {
  return JSON.stringify(path.map((segment) => {
    return typeof segment === "bigint"
      ? ["bigint", String(segment)]
      : [typeof segment, segment];
  }));
}

function validateEntry(
  source: string,
  resourceType: string,
  mode: PolicyMode,
  entry: unknown,
): string {
  const context = `${source} ${mode} entry for ${resourceType}`;
  if (!isRecord(entry)) {
    return fail(`${context} must be an object`);
  }
  rejectUnknownKeys(entry, entryKeys(mode), context);
  for (const key of requiredStrings(mode)) {
    requireString(entry, key, context);
  }
  if (Object.hasOwn(entry, "ticket")) {
    const ticket = entry.ticket;
    if (typeof ticket !== "string" || ticket.length === 0) {
      fail(`${context} has invalid ticket`);
    }
  }

  if (mode === "projection_sync") {
    const targetText = requireString(entry, "target_path", context);
    const sourceText = requireString(entry, "source_path", context);
    const target = policyPath(targetText);
    const sourcePath = policyPath(sourceText);
    if (policyPathsEqual(target, sourcePath)) {
      fail(`${source} projection_sync entry for ${resourceType} target_path and source_path must differ`);
    }
    if (policyPathHasWildcardOrIndex(target)) {
      fail(`${source} projection_sync entry for ${resourceType} target_path must not contain wildcard or index selectors`);
    }
    if (policyPathHasWildcardOrIndex(sourcePath)) {
      fail(`${source} projection_sync entry for ${resourceType} source_path must not contain wildcard or index selectors`);
    }
    return `projection_sync\0${targetText}`;
  }

  const pathText = requireString(entry, "path", context);
  const parsed = policyPath(pathText);
  if (mode === "projection_fill") {
    const sourceText = requireString(entry, "source", context);
    const sourcePath = policyPath(sourceText);
    if (parsed.length !== 1) {
      fail(`${source} projection_fill entry for ${resourceType} path must be a single top-level name`);
    }
    if (sourcePath.length !== 1) {
      fail(`${source} projection_fill entry for ${resourceType} source must be a single top-level raw API name`);
    }
    if (policyPathHasWildcardOrIndex(parsed)) {
      fail(`${source} projection_fill entry for ${resourceType} path must not contain wildcard or index selectors`);
    }
    if (policyPathHasWildcardOrIndex(sourcePath)) {
      fail(`${source} projection_fill entry for ${resourceType} source must not contain wildcard or index selectors`);
    }
    return `projection_fill\0${pathText}`;
  }
  if (mode === "projection_omit_if") {
    const values = entry.values;
    if (!Array.isArray(values) || values.length === 0) {
      fail(`${source} projection_omit_if entry for ${resourceType} values must be a non-empty JSON list`);
    }
    if (!values.every(isJsonScalar)) {
      fail(`${source} projection_omit_if entry for ${resourceType} values must contain only JSON scalars`);
    }
    return `projection_omit_if\0${pathText}\0${JSON.stringify(values.map(jsonScalarMarker))}`;
  }
  if (mode === "plan_tolerate") {
    const rawActions = Object.hasOwn(entry, "actions")
      ? entry.actions
      : ["update"];
    if (!Array.isArray(rawActions)) {
      fail(`${source} plan_tolerate entries for ${resourceType} actions must be a list`);
    }
    if (rawActions.length === 0) {
      fail(`${source} plan_tolerate entry for ${resourceType} actions must not be empty`);
    }
    const seen = new Set<string>();
    for (const action of rawActions) {
      if (typeof action !== "string" || action.length === 0) {
        fail(`${source} plan_tolerate entry for ${resourceType} has invalid action`);
      }
      if (action !== "update") {
        fail(`${source} plan_tolerate entry for ${resourceType} has unsupported action ${JSON.stringify(action)}`);
      }
      if (seen.has(action)) {
        fail(`${source} plan_tolerate entry for ${resourceType} has duplicate action ${JSON.stringify(action)}`);
      }
      seen.add(action);
    }
    return `plan_tolerate\0${pathText}\0${sortedStrings(seen).join("\0")}`;
  }
  return `projection_omit\0${pathText}\0${pathMarker(parsed)}`;
}

function validatePolicy(data: unknown, source: string): Record<string, unknown> {
  if (!isRecord(data)) {
    return fail(`${source}: drift policy must be an object`);
  }
  rejectUnknownKeys(data, TOP_LEVEL_KEYS, `${source} top-level drift policy`);
  if (!Object.hasOwn(data, "version")) {
    fail(`${source}: drift policy missing version`);
  }
  if (!isSupportedDriftPolicyVersion(data.version)) {
    fail(`${source}: unsupported drift policy version`);
  }
  if (!Object.hasOwn(data, "resource_types")) {
    fail(`${source}: drift policy missing resource_types`);
  }
  if (!isRecord(data.resource_types)) {
    fail(`${source}: resource_types must be an object`);
  }
  let entryCount = 0;
  for (const resourceType of sortedStrings(Object.keys(data.resource_types))) {
    if (!/^[A-Za-z_][A-Za-z0-9_]*$/.test(resourceType)) {
      fail(`${source}: invalid resource type ${JSON.stringify(resourceType)}`);
    }
    const resource = data.resource_types[resourceType];
    if (!isRecord(resource)) {
      fail(`${source}: policy for ${resourceType} must be an object`);
    }
    rejectUnknownKeys(resource, RESOURCE_KEYS, `${source} policy for ${resourceType}`);
    for (const mode of MODES) {
      const rawEntries = Object.hasOwn(resource, mode) ? resource[mode] : [];
      if (!Array.isArray(rawEntries)) {
        fail(`${source} ${mode} entries for ${resourceType} must be a list`);
      }
      entryCount += rawEntries.length;
      if (entryCount > MAX_POLICY_ENTRIES) {
        fail(`${source}: drift policy exceeds the entry-count limit`);
      }
      const scopes = new Set<string>();
      for (const entry of rawEntries) {
        const scope = validateEntry(source, resourceType, mode, entry);
        if (scopes.has(scope)) {
          const display = isRecord(entry)
            ? entry.path ?? entry.target_path
            : "unknown";
          fail(`${source} duplicate ${mode} entry for ${resourceType} path ${String(display)}`);
        }
        scopes.add(scope);
      }
    }
    const fill = new Map<string, string>();
    for (const entry of entriesFor(data, resourceType, "projection_fill")) {
      const text = entry.path as string;
      fill.set(pathMarker(policyPath(text)), text);
    }
    for (const entry of entriesFor(data, resourceType, "projection_omit")) {
      const text = entry.path as string;
      if (fill.has(pathMarker(policyPath(text)))) {
        fail(`${source} projection_fill and projection_omit entries for ${resourceType} conflict on path ${text}`);
      }
    }
  }
  return data;
}

export interface StalePolicyEntry {
  readonly resource_type: string;
  readonly mode: PolicyMode;
  readonly path: string;
}

export class DriftPolicy {
  readonly data: Record<string, unknown>;
  readonly source: string;
  private readonly matched = new WeakSet<object>();
  private readonly exactPlanTolerate = new Map<
    string,
    ReadonlyMap<string, CompiledPlanTolerate>
  >();
  private readonly wildcardPlanTolerate = new Map<
    string,
    readonly CompiledPlanTolerate[]
  >();

  constructor(data: unknown = null, source = "<memory>") {
    this.source = source;
    this.data = validatePolicy(
      data === null ? { version: 1, resource_types: {} } : data,
      source,
    );
    const resources = this.data.resource_types as Record<string, unknown>;
    for (const resourceType of Object.keys(resources)) {
      const exact = new Map<string, CompiledPlanTolerate>();
      const wildcard: CompiledPlanTolerate[] = [];
      for (const [order, entry] of this.entries(
        resourceType,
        "plan_tolerate",
      ).entries()) {
        const selector = policyPath(entry.path);
        const compiled = { entry, order, selector };
        if (selector.includes(POLICY_WILDCARD)) {
          wildcard.push(compiled);
        } else if (!exact.has(pathMarker(selector))) {
          // Python walks entries in declaration order. Textually distinct
          // selectors such as field[0] and field[00] can canonicalize to the
          // same path, so retain the first and leave later aliases stale.
          exact.set(pathMarker(selector), compiled);
        }
      }
      if (wildcard.length > MAX_PLAN_TOLERATE_WILDCARDS_PER_RESOURCE) {
        fail(`${source}: plan_tolerate wildcard entries exceed the per-resource limit`);
      }
      this.exactPlanTolerate.set(resourceType, exact);
      this.wildcardPlanTolerate.set(resourceType, wildcard);
    }
  }

  entries(resourceType: string, mode: PolicyMode): PolicyEntry[] {
    return [...entriesFor(this.data, resourceType, mode)];
  }

  markMatched(entry: PolicyEntry): void {
    this.matched.add(entry);
  }

  projectionOmits(
    resourceType: string,
    path: readonly ConcretePathSegment[],
  ): boolean {
    for (const entry of this.entries(resourceType, "projection_omit")) {
      if (policySelectorMatches(policyPath(entry.path), path)) {
        this.matched.add(entry);
        return true;
      }
    }
    return false;
  }

  toleratesPlanPath(
    resourceType: string,
    path: readonly ConcretePathSegment[],
    action: string,
  ): boolean {
    if (action !== "update") {
      return false;
    }
    const candidates: CompiledPlanTolerate[] = [];
    const exact = this.exactPlanTolerate.get(resourceType)?.get(pathMarker(path));
    if (exact !== undefined && policySelectorMatches(exact.selector, path)) {
      candidates.push(exact);
    }
    for (const candidate of this.wildcardPlanTolerate.get(resourceType) ?? []) {
      if (policySelectorMatches(candidate.selector, path)) {
        candidates.push(candidate);
      }
    }
    candidates.sort((left, right) => left.order - right.order);
    const matched = candidates[0];
    if (matched !== undefined) {
      this.matched.add(matched.entry);
      return true;
    }
    return false;
  }

  staleEntries(options: {
    resourceTypes?: ReadonlySet<string>;
    modes?: readonly PolicyMode[];
  } = {}): StalePolicyEntry[] {
    const modes = options.modes === undefined || options.modes.length === 0
      ? MODES
      : options.modes;
    const stale: StalePolicyEntry[] = [];
    const resources = this.data.resource_types as Record<string, unknown>;
    for (const resourceType of sortedStrings(Object.keys(resources))) {
      if (
        options.resourceTypes !== undefined
        && options.resourceTypes.size > 0
        && !options.resourceTypes.has(resourceType)
      ) {
        continue;
      }
      for (const mode of modes) {
        for (const entry of this.entries(resourceType, mode)) {
          if (!this.matched.has(entry)) {
            stale.push({
              resource_type: resourceType,
              mode,
              path: String(entry.path ?? entry.target_path),
            });
          }
        }
      }
    }
    return stale;
  }
}
